packages/store/addon/-private/system/record-data-for.ts

```diff
@@ -19,11 +22,26 @@ interface InternalModel {
 type DSModelOrSnapshot = { _internalModel: InternalModel };
 type Reference = { internalModel: InternalModel };
 
-type Instance = InternalModel | RecordData | DSModelOrSnapshot | Reference;
+type Instance = StableRecordIdentifier | InternalModel | RecordData | DSModelOrSnapshot | Reference;
+
+const IdentifierCache = new WeakMap<StableRecordIdentifier, RecordData>();
+
+export function setRecordDataFor(identifier: StableRecordIdentifier, recordData: RecordData) {
+  assert(`Illegal set of identifier`, !IdentifierCache.has(identifier));
+  IdentifierCache.set(identifier, recordData);
+}
+
+export function removeRecordDataFor(identifier) {
+  IdentifierCache.delete(identifier);
+}
 
+export default function recordDataFor(instance: StableRecordIdentifier): RecordData | null;
 export default function recordDataFor(instance: Instance): RecordData;
 export default function recordDataFor(instance: object): null;
 export default function recordDataFor(instance: Instance | object): RecordData | null {
+  if (IdentifierCache.has(instance as StableRecordIdentifier)) {
+    return IdentifierCache.get(instance as StableRecordIdentifier) as RecordData;
```

Comment:
I'm surprised `as RecordData` cast is necessary.  I would have thought TS would have seen through to the return type.

Reply:
`has` doesn't narrow is why, so TS sees it as `| undefined`. It will narrow using the `get` first and then checking for `undefined` instead of using `has`, which is the pattern we use when we may also need to assign.